packages/ingest/recursion.js

```diff
@@ -1,84 +1,93 @@
 
-const join = require('path').join;
-const log = require('@cumulus/common/log');
-
 /**
- * Handles recursion of a FTP/SFTP list operation
- * It requests a promisified list function that returns contents of
- * a directory on a server
+ * check if segment is a regex segment
  *
- * @param {function} fn The promisified function for listing a directory
- * @param {string} originalPath the full path to use for recursively listing the directory
- * @param {string} [currentPath=null] the current directory to list recursively
- * @param {number} [position=0] current position in the recursive tree
- * @returns {Promise} the promise of an object that has the path is the key and
- *   list of files as values
+ * @param {string} segment - path segment
+ * @returns {boolean}
  */
-async function recursion(fn, originalPath, currentPath = null, position = 0) {
-  // build the recursion path object
-  const regex = /(\(.*?\))/g;
-  const rules = (originalPath || '/').split(regex).map((i) => i.replace(/\\\\/g, '\\'));
-  const map = rules.map((r) => (r.match(regex) !== null));
-
-  let files = [];
-  let path = currentPath;
-  if (!path) {
-    path = rules[position];
-  }
+const isRegexSegment = (segment) => segment.match(/(\([^\)]*\))/g) !== null;
 
-  log.info(`Listing ${path}`);
-
-  // get list of current path
-  const list = await fn(path);
-
-  // loop try what is returned
-  for (let ctr = 0; ctr < list.length; ctr += 1) {
-    const item = list[ctr];
+/**
+ * Insert leading and terminating slashes into the path string if not present
+ *
+ * @param {string} path - path string
+ * @returns {string} normalized path
+ */
+const normalizeWithSlashes = (path) => {
+  let output = path;
+  if (!path.startsWith('/')) output = `/${output}`;
+  if (!path.endsWith('/')) output = `${output}/`;
+  return output;
+};
 
-    let regexPath;
-    let textPath;
-    let newPath;
+/**
+ * Recur on directory, list all files, and recur into any further directories,
+ * as specified regex segments allow.
+ * Note that the list function will be called with a path argument that includes
+ * both leading and terminating slashes. List functions will need to be able to handle
+ * or remove leading and terminating slashes accordingly.
+ *
+ * @param {Function} fn - list function
+ * @param {string} currentPath - current path to list
+ * @param {Array<string>} segments - path segments
+ * @param {number} position - current position in the segment list
+ */
+async function recurOnDirectory(fn, currentPath, segments, position) {
+  // check if we have a filter regex segment (e.g. '(dir.*)')
+  const filterExpr = segments[position + 1] || '';
+  const isRegex = isRegexSegment(filterExpr);
 
-    // if directory is found, apply recursion logic
-    if (item.type === 'd' || item.type === 1) {
-      // first we check if the next segment of the provided path
-      // is a regex rule
-      const isRegex = map[position + 1];
+  const path = currentPath.replace(/\/+/g, '/');
```

Comment:
Replacing multiple trailing slashes with a single trailing slash?

Reply:
I'm being a bit pedantic on this point, but this would also replace single slashes with single slashes, which is obviously pointless. Therefore, if you feel like getting pedantic with me, you should use `/[/]{2,}/g`. (Note: I tend to prefer using `[]` rather than `\`, which is arguably slightly more readable in many cases.)

Reply:
We need to replace leading slashes, as it is possible to just be listing root (i.e. `/`), then when some directory is discovered the next path is `${path}/${item.name}/`. When `path` is `/`, this results in `//someDirectory/`, which may or may not be a valid path. I figured it was safest to just standardize it to remove all double slashes, leading, trailing, otherwise.